### PR TITLE
build/ubuntu: improve build-repo.sh for incremental builds

### PR DESCRIPTION
**Pull request by jialeif**
In case of a build failure, re-running build-repo.sh can skip the same and continue the build where it failed.

### CODE REVIEW

**Review by kenplusplus — COMMENTED** [review]
If I want to do a clean build, do I need delete build.done from each directory in 1~2 minutes?

**Comment by jialeif** [comment]
> If I want to do a clean build, do I need delete build.done from each directory in 1~2 minutes?

 In which case do you need to delete all build.done. `git clean -fd` is clean way, it will delete more obj files.

**Comment by kenplusplus** [comment]
> > If I want to do a clean build, do I need delete build.done from each directory in 1~2 minutes?
> 
> In which case do you need to delete all build.done. `git clean -fd` is clean way, it will delete more obj files.

So if I made some changes, it will be also clean, we should not leverage git to delete build.done.

**Comment by jialeif** [comment]
> > > If I want to do a clean build, do I need delete build.done from each directory in 1~2 minutes?
> > 
> > 
> > In which case do you need to delete all build.done. `git clean -fd` is clean way, it will delete more obj files.
> 
> So if I made some changes, it will be also clean, we should not leverage git to delete build.done.

Yes, this is one case, we need delete the build.done in the modified components.

**Review by kenplusplus — CHANGES_REQUESTED** [review]
So please provides an easier way to cleanup build.done

**Comment by jialeif** [comment]
> So please provides an easier way to cleanup build.done

In the case of modify only one component, it doesn't make sense to clean all build.done.

**Comment by kenplusplus** [comment]
> > So please provides an easier way to cleanup build.done
> 
> In the case of modify only one component, it doesn't make sense to clean all build.done.

the requirement is to provide an approach to clean all build.done. I am not argue make sure or not, please do it now

**Comment by HaokunX-intel** [comment]
The build caches can be placed in one place, such as the directory where the `build/ubuntu-22.04/build-repo.sh` resides and is named by `build/ubuntu-22.04/build-cache`. `touch` a flag file for each component with different name, such as `build/ubuntu-22.04/build-cache/libvirt.done` and `build/ubuntu-22.04/build-cache/qemu.done`. If all building pass, we can delete the directory `build-cache`.

**Comment by kenplusplus** [comment]
> The build caches can be placed in one place, such as the directory where the `build/ubuntu-22.04/build-repo.sh` resides and is named by `build/ubuntu-22.04/build-cache`. `touch` a flag file for each component with different name, such as `build/ubuntu-22.04/build-cache/libvirt.done` and `build/ubuntu-22.04/build-cache/qemu.done`. If all building pass, we can delete the directory `build-cache`.

This is very nice design and idea! Thanks Haokun!

**Comment by jialeif** [comment]
> > The build caches can be placed in one place, such as the directory where the `build/ubuntu-22.04/build-repo.sh` resides and is named by `build/ubuntu-22.04/build-cache`. `touch` a flag file for each component with different name, such as `build/ubuntu-22.04/build-cache/libvirt.done` and `build/ubuntu-22.04/build-cache/qemu.done`. If all building pass, we can delete the directory `build-cache`.
> 
> This is very nice design and idea! Thanks Haokun!

Nice! How about build-status?

**Comment by intelzhongjie** [comment]
Here are my thoughts FYI. Point out if my understanding is wrong:
1. It may be better that adding an option for build-repo.sh so caller can decide if it's "incremental build" or "clean build from scratch"
2. Please consider this case: first, build-repo.sh start, then one of the "sub-build" failed but others succeeded, just at this time, someone's commit to one of the repos got merged which built successfully just in previous attempt, then is your current logic able to handle this case correctly as required?

**Comment by jialeif** [comment]
> Here are my thoughts FYI. Point out if my understanding is wrong:
> 
> 1. It may be better that adding an option for build-repo.sh so caller can decide if it's "incremental build" or "clean build from scratch". Is there any existing "cache" mechanism of make/compile used by our case?
> 2. Please consider this case: first, build-repo.sh start, then one of the "sub-build" failed but others succeeded, just at this time, someone's commit to one of the repos got merged which built successfully just in previous attempt, then is your current logic able to handle this case correctly as required?

Thanks all! Let me update it.

**Comment by jialeif** [comment]
@kenplusplus @HaokunX-intel @intelzhongjie  Please review the changes.

File: build/ubuntu-22.04/build-repo.sh [hunk 1]
```diff
@@ -3,68 +3,79 @@
 THIS_DIR=$(dirname "$(readlink -f "$0")")
 GUEST_REPO="guest_repo"
 HOST_REPO="host_repo"
+STATUS_DIR="${THIS_DIR}/build-status"
 
 export DEBIAN_FRONTEND=noninteractive
 
-build_check() {
+build_check( ) {
```
Comment (intelzhongjie) [inline build/ubuntu-22.04/build-repo.sh:10]: Just remove this unnecessary space here

File: build/ubuntu-22.04/build-repo.sh [hunk 2]
```diff
@@ -3,68 +3,79 @@
...
     if [[ $(id -u) -eq 0 ]]; then
         echo "Running the whole script as root is not recommended. virnetsockettest might be failed when building libvirt as root"
         echo "But mk-build-deps needs sudo to install build dependecies, we should setup passwordless sudo for automation"
-	exit 1
+        exit 1
+    fi
+
+    [[ -d $STATUS_DIR ]] || mkdir $STATUS_DIR
+    if [[ "$1" == clean-build ]]; then
```
Comment (intelzhongjie) [inline build/ubuntu-22.04/build-repo.sh:18]: So the current logic implies that the default option is "incremental", correct?

File: build/ubuntu-22.04/build-repo.sh [hunk 3]
```diff
@@ -3,68 +3,79 @@
...
-	    libvirt-daemon-driver-qemu_*.deb libvirt-daemon-config-network_*.deb libvirt-daemon-config-nwfilter_*.deb \
-	    libvirt-login-shell_*.deb libvirt-daemon-driver-lxc_*.deb ../$HOST_REPO/
-    cd ..
+            libvirt-daemon-driver-qemu_*.deb libvirt-daemon-config-network_*.deb libvirt-daemon-config-nwfilter_*.deb \
+            libvirt-login-shell_*.deb libvirt-daemon-driver-lxc_*.deb ../$HOST_REPO/
+    popd
 }
 
-build_check
+build_check $1
```
Comment (intelzhongjie) [inline build/ubuntu-22.04/build-repo.sh:78]: Then please carefully consider the usage of this $1 in the caller and the default behavior when $1 is none/empty.

File: build/ubuntu-22.04/build-repo.sh (hunk 1 above)
Comment (jialeif) [reply build/ubuntu-22.04/build-repo.sh]: updated

File: build/ubuntu-22.04/build-repo.sh (hunk 3 above)
Comment (jialeif) [reply build/ubuntu-22.04/build-repo.sh:78]: default for incremental build

**Review by intelzhongjie — APPROVED** [review]
ship it!

**Review by kenplusplus — APPROVED** [review]
LGTM

**Review by kenplusplus — APPROVED** [review]
LGTM

**Review by kenplusplus — APPROVED** [review]
LGTM

**Review by kenplusplus — APPROVED** [review]
LGTM

**Pull request merged**